src/pycountry/__init__.py

```diff
@@ -126,101 +124,125 @@ class HistoricCountries(ExistingCountries):
-            if parts[0] != self.country_code:
-                self.parent_code = f"{self.country_code}-{self.parent_code}"
-
     @property
-    def country(self):
-        return countries.get(alpha_2=self.country_code)
+    def country(self) -> pycountry.db.Country:
+        return cast(
+            pycountry.db.Country, countries.get(alpha_2=self.country_code)
+        )
+
+    @cached_property
+    def country_code(self) -> str:
+        return self.code.split("-")[0]
 
     @property
-    def parent(self):
+    def parent(self) -> Optional["SubdivisionHierarchy"]:
         if not self.parent_code:
             return None
         return subdivisions.get(code=self.parent_code)
 
+    @cached_property
+    def parent_code(self) -> Optional[str]:
+        parent = self._fields.get("parent")
+        if parent is not None:
+            # check if the country_code is already present
+            parts = parent.split("-")
+            if parts[0] != self.country_code:
+                parent = f"{self.country_code}-{parent}"
+
+        return parent
+
 
-class Subdivisions(pycountry.db.Database):
+class Subdivisions(pycountry.db.Database[SubdivisionHierarchy]):
     # Note: subdivisions can be hierarchical to other subdivisions. The
     # parent_code attribute is related to other subdivisions, *not*
     # the country!
 
-    data_class = SubdivisionHierarchy
+    factory = SubdivisionHierarchy
     no_index = ["name", "parent_code", "parent", "type"]
     root_key = "3166-2"
 
-    def _load(self, *args, **kw):
-        super()._load(*args, **kw)
+    def _load(self) -> None:
+        super()._load()
 
         # Add index for the country code.
         self.indices["country_code"] = {}
         for subdivision in self:
-            divs = self.indices["country_code"].setdefault(
-                subdivision.country_code.lower(), set()
+            divs = cast(
+                Set[SubdivisionHierarchy],
+                self.indices["country_code"].setdefault(
+                    subdivision.country_code.lower(), set()  # type: ignore[arg-type]
+                ),
             )
             divs.add(subdivision)
 
-    def get(self, **kw):
-        default = kw.setdefault("default", None)
-        subdivisions = super().get(**kw)
-        if subdivisions is default and "country_code" in kw:
+    def get(
+        self, *, default: Optional[SubdivisionHierarchy] = None, **kw: str
+    ) -> Optional[SubdivisionHierarchy]:
+        result = super().get(default=default, **kw)
+        if result is default and "country_code" in kw:
             # This handles the case where we know about a country but there
             # are no subdivisions: we return an empty list in this case
             # (sticking to the expected type here) instead of None.
             if countries.get(alpha_2=kw["country_code"]) is not None:
-                return []
-        return subdivisions
+                return []  # type: ignore[return-value]
```

Comment:
I'm a little confused why this is an empty list since `super().get(...)` will return an `Optional[SubdivisionHierarchy]` and this would have to be typed `| list[]`. For now I ignored the return type, but if this is correct the method signature should be updated.